sgx demo initial remote attestation

lkatalin opened the pull request:
I probably want to make some more changes, ex. pass the certs by command line arg instead of including the files, but just so you have this to look at ASAP.

npmccallum commented on intel-sgx/attestation-enclave/src/main.rs:
```diff
@@ -0,0 +1,58 @@
+    println!("\nListening on port 1032....\n");
+
+    for stream in TcpListener::bind("localhost:1032").unwrap().incoming() {
+        match stream {
```
Don't use match here (or in other files either). This is just a demo. Let's `unwrap()` it for readability.

lkatalin replied:
I removed all the matches and now have unwrap() or sometimes expect() in case error messages seemed useful.

npmccallum commented on intel-sgx/attestation-enclave/src/main.rs:
```diff
@@ -0,0 +1,58 @@
+
+fn create_report(t: Targetinfo) -> Report {
+    Report::for_target(&t, &[0; 64])
+}
```
Don't use micro functions like this (fix other places too). It hurts readability. If a function is less than a handful of lines, just do it inline. I'd much rather see:

```rust
// Other code ...

// Create report
let report = Report::for_target(&t, &[0; 64]);

// Other code ...
```

lkatalin replied:
Got it, I put these short lines of code back in main() for all the files. For the tenant's verification of the Quote, I still broke it down into individual functions because to me that seemed more clear, but let me know.

steveeJ replied:
I like to use inline scopes for separating functionality out without creating an actual function. That works nicely with scopes which just return one value.

lkatalin commented on intel-sgx/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "sgx-demo"
+version = "0.1.0"
+authors = ["LKatalin"]
```
Done.

npmccallum replied:
I don't think this is fixed.

lkatalin commented on intel-sgx/attestation-daemon/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "attestation-daemon"
+version = "0.1.0"
+authors = ["LKatalin"]
```
Done.

lkatalin commented on intel-sgx/attestation-enclave/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "attestation-enclave"
+version = "0.1.0"
+authors = ["LKatalin"]
```
Done.

lkatalin commented:
@npmccallum Please take a look at the changes when you get a chance. Thanks for pointing out that the tenant needs to do verification; I strongly agree. I initially meant to do it that way but was blocked by something (now I don't recall what that was and it doesn't matter because I fixed it). My current next step is also to get the cert chain loaded in a more intelligent way.

npmccallum commented on README.md:
```diff
@@ -1,8 +1,27 @@
+2. Have Rust Nightly and the Fortanix EDP installed, following the steps on this [page](https://github.com/fortanix/rust-sgx). 
+3. Retrieve Intel's PCK certificate chain as described in the section above.
+4. Run the `attestation-enclave` with `cargo run --target x86_64-fortanix-unknown-sgx` and leave it running. Run the `attestation-daemon` with `cargo run` and leave it running.
+5. Run the `attestation-tenant` with `cargo run <filepath>`, where filepath is the path to the PCK certificate chain from Step 3.
```
Please wrap the `README.md` at 80 columns. Thanks!

lkatalin replied:
I wrapped it at 80 columns. It may be a little too short now; please let me know if so.

npmccallum commented on intel-sgx/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+
+[dependencies]
+sgx-isa = "0.3.1"
+dcap-ql = "0.2.0"
```
What is the purpose of these dependencies? This top-level directory should just be an empty workspace.

lkatalin replied:
I've removed these from the top level.

npmccallum commented on intel-sgx/attestation-daemon/src/main.rs:
```diff
@@ -0,0 +1,33 @@
+extern crate dcap_ql;
```
`extern crate CRATE;` is no longer necessary in stable. Please remove it.

lkatalin replied:
Done.

npmccallum commented on intel-sgx/attestation-daemon/src/main.rs:
```diff
@@ -0,0 +1,33 @@
+        let mut stream_client = stream_client.unwrap();
+
+        // connect to enclave
+        let mut enclave_cnx = TcpStream::connect("localhost:1032").unwrap();
```
This is just a comment. No action is required. It is really gross that we have to hard-code the address and port here. In the future, the SGX enclave management daemon (probably named `enarx-sgx`) should spawn the enclaves directly. Is it possible to pass a `socketpair()` to the enclave for management communication? We should probably discuss this at some point.

lkatalin replied:
I'd be happy to discuss this and create this in a cleaner way. I can open an issue for it, also.

npmccallum commented on intel-sgx/attestation-daemon/src/main.rs:
```diff
@@ -0,0 +1,33 @@
+        println!("\nQuote successfully generated...");
+
+        // send quote to tenant
+        stream_client.write(&quote).unwrap();
```
I understand what each of these chunks of code does. But I don't understand *why* this code works like this. Could you add additional information to the comments explaining this? Also, be sure that comments are complete sentences with proper grammar and punctuation.

lkatalin replied:
Yes, that's a good point, and I added additional comments to each part of the code to explain its purpose (ex. that the Targetinfo is needed to generate the Report). But please let me know if anything is still unclear.

npmccallum commented on intel-sgx/attestation-enclave/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+edition = "2018"
+
+[dependencies]
+sgx-isa = { version = "0.3.1", features = ["sgxstd"] }
```
Please document what this feature does and why you turn it on.

lkatalin replied:
Done.

npmccallum commented on intel-sgx/src/main.rs:
```diff
@@ -0,0 +1 @@
+fn main() {}
```
This file should be unnecessary.

lkatalin replied:
I removed this.

npmccallum commented:
Also, I merged @mbestavros's PR, so now there is a conflict. Please resolve it.

lkatalin commented:
> 1. Comments need to be complete sentences with proper grammar and punctuation.
I've added full-sentence comments that are more descriptive. Please let me know if any are still unclear!

> 2. Is there a reason the enclave can't just send the report to dcap directly and then forward the completed quote to the daemon?
Yes, the reason is that currently the `dcap-ql` crate can't be used in an enclave. We will need to write our own `enclave-runner` to add this functionality. I could open an issue for this as well.

> 3. We really should avoid shipping raw structs over the wire. Can the `Quote` type be used with Serde for serialization?
I've added Serde support for the Quote, Targetinfo, and Report types in a [PR to Fortanix](https://github.com/fortanix/rust-sgx/pull/177). Currently the Targetinfo and Report types are serialized into strings with `serde_json` and then sent as bytes. If this doesn't make sense, I will do it a different way. `serde_json` also has a `vec` type, but it requires a JSON map. The Quote was and still is sent over the wire as a `Vec<u8>`, not as a struct, so that is not serialized.

> 4. The same is true for parsing inside the tenant. We really don't want to be doing byte-level parsing. There should be pre-defined data types that use serialization (see my comment about Serde above).
I also opened a [PR to Fortanix](https://github.com/fortanix/rust-sgx/pull/173) to have support for Quote-related types for PCK cert chain-based attestation, and currently the `attestation-tenant` makes use of these. 

> 5. OpenSSL (and mbedtls) should have functions for the type conversions you're doing. Please confirm.
I've removed the manual conversions (to ASN.1, for example) and am using the OpenSSL versions now.

> 6. It is unclear to me that `verify_chain_sigs()` actually does signature validation. At least that's not how I'm used to doing signature validation. Traditionally, I've built up a full certificate chain and called a single verify function on the whole chain. For example, if Intel decided to add multiple intermediate certificates this could would have to be re-written.
I've changed this to build up the entire chain and then verify only the leaf PCK cert in the context of the chain. Hopefully this is what you meant.

> 7. Right now the tenant code is just a huge pile of functions. I'd prefer if there was discrete data types.
I grouped the functions together and came up with `chain`, `key`, and `sig` types that each have their own file and are imported as modules. These types have the old functions as methods.

Other thoughts:

- In addition to the separate PRs submitted to Fortanix for `serde` and `PCK cert chain`-related struct support, I created a [third branch](https://github.com/lkatalin/rust-sgx/tree/serde-pck) of my `rust-sgx` fork, which merges both of these functionalities. Currently the `intel-sgx` `Cargo.toml` depends on this third branch of the `rust-sgx` fork. When both PRs are merged to Fortanix, the dependency can point to Fortanix's repo.

- I don't like the way the size of the buffers receiving the serialized structs is hardcoded right now, as it's messy and pretty brittle. I can open an issue to change this too, but I need to look into a better way to do this.

steveeJ commented on intel-sgx/attestation-daemon/src/main.rs:
```diff
@@ -0,0 +1,44 @@
+            .read_exact(&mut report_buf)
+            .expect("Could not read report");
+        let serialized_report = String::from_utf8(report_buf.to_vec()).unwrap();
+        let report: sgx_isa::Report = serde_json::from_str(&serialized_report).unwrap();
```
I think you could use a conjunction of https://doc.rust-lang.org/std/io/trait.Read.html#method.take and https://docs.serde.rs/serde_json/de/fn.from_reader.html to skip the buffer and the string conversion altogether.

lkatalin replied:
I made this change in this file, but I don't see a way to do it in `attestation-enclave/src/main.rs` because `from_reader(stream)` consumes my stream, and I need to write to that same stream later. Any thoughts?

steveeJ commented on README.md:
```diff
@@ -9,3 +9,6 @@ Compiling either a C or Rust program to WASM and running it natively using a Rus
 Running native Rust code in a WASM virtual machine without pre-compiling it.
+
+## Intel SGX
+A demonstration of remote attestation for an SGX enclave.
```
Maybe you could elaborate a bit here, giving a high-level description for the demo flow.

steveeJ commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,50 @@
+        let xbn = BigNum::from_slice(&x).unwrap();
+        let ybn = BigNum::from_slice(&y).unwrap();
+        let ec_key = EcKey::from_public_key_affine_coordinates(&group, &xbn, &ybn).unwrap();
+        assert!(ec_key.check_key().is_ok());
```
I find it surprising to see an assertion here. Could you instead return a `Result<Self>`, remove this line and use `?` in the previous line instead of unwrapping? The same may be true for all other unwraps and assertions in this file. Then again, I'm not sure serious this code is so feel free to ignore my comment completely.

steveeJ commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+impl Key {
+    pub fn new_from_xy(xy_coords: &[u8]) -> Self {
```
I'd expect a `///` style comment on top of any `pub` keyword

steveeJ commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,50 @@
+};
+
+//#[derive(Borrow)]
+pub struct Key {
```
I'd expect a `///` style comment on top of any `pub` keyword

steveeJ commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,50 @@
+        Key { pkey: pkey }
+    }
+
+    pub fn new_from_pubkey(pkey: PKey<Public>) -> Self {
```
I'd expect a `///` style comment on top of any `pub` keyword

steveeJ commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,50 @@
+        Key { pkey: pkey }
+    }
+
+    pub fn verify_sig(&self, signed: &[u8], sig: &Vec<u8>) -> () {
```
I'd expect a `///` style comment on top of any `pub` keyword

steveeJ commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,50 @@
+        assert!(verifier.verify(sig).unwrap());
+    }
+
+    // This verifies the SHA-256 hash of the Attestation Public Key || QEAuthData
```
Please turn this comment into a `///` prefixed one. Thanks for describing what happens here. I'm curious to learn about the cryptography from these comments.

steveeJ commented on intel-sgx/attestation-tenant/src/cert_chain.rs:
```diff
@@ -0,0 +1,74 @@
+        // This operation verifies the leaf (PCK_cert) in the context of the
+        // chain. If the chain cannot be verified, the leaf will not be
+        // verified.
+        assert!(context
```
maybe return a `Result<()>` instead the assertion here. see similar comments on other files

steveeJ commented on intel-sgx/attestation-tenant/src/main.rs:
```diff
@@ -0,0 +1,113 @@
+};
+
+const DAEMON_CONN: &'static str = "localhost:1034";
+const VEC_QUOTE_SIZE: usize = 4702;
```
that's an untypical number. do you mind sharing how this size is determined, e.g. with a reference?

lkatalin replied:
Looks like I don't need to use this size at all, thankfully.

steveeJ commented on intel-sgx/attestation-tenant/src/main.rs:
```diff
@@ -0,0 +1,113 @@
+// The tenant requests attestation of an enclave from the platform's attestation daemon, and
+// receives a Quote from the daemon. The Quote verifies the enclave's measurement. The tenant
+// verifies:
+// 1. That the Quote's PCK Certificate (embedded in the Cert Data) is valid.
```
Maybe you could refer to the spec which describes _PCK_ and other acronyms somewhere in the top of this file?

lkatalin replied:
I added two links to the top of the file. However, there isn't any very good, concise, and up to date documentation that I can find... hoping to put something together in the near future.

steveeJ commented on intel-sgx/attestation-tenant/src/main.rs:
```diff
@@ -0,0 +1,113 @@
+// 3. That the Attestation Key signed the Quote.
+// 4. That the hashed material (containing the Attestation Key) signed by the PCK is valid.
+
+fn main() {
```
Personally I would probably refactor this main and extract some functions with descriptive names, but feel free to ignore this suggestion.

lkatalin replied:
I would like to get @npmccallum 's input before I do that, as earlier I had refactored to have more code in main rather than extracted functions which were very short.

steveeJ replied:
Yep, I saw we have different opinions here. My isn't very strong though, and I suggest to use scopes as a compromise in the future.

steveeJ commented on intel-sgx/attestation-tenant/src/sig.rs:
```diff
@@ -0,0 +1,80 @@
+use std::convert::TryFrom;
+
+#[derive(Copy, Clone)]
+pub struct Signature {
```
missing `///`on a `pub`

lkatalin replied:
What do you mean by this one? I don't think I'm familiar.

steveeJ commented on intel-sgx/attestation-tenant/src/sig.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+impl Signature {
+    pub fn to_der_vec(self) -> Vec<u8> {
```
missing `///`on a `pub`

steveeJ commented on intel-sgx/attestation-tenant/src/cert_chain.rs:
```diff
@@ -0,0 +1,74 @@
+#[allow(dead_code)]
+impl CertChain {
+    // The chain does not include the leaf.
+    pub fn new_from_chain(c: Vec<X509>, leaf: &X509) -> Self {
```
missing `///` on `pub`

steveeJ commented on intel-sgx/attestation-tenant/src/cert_chain.rs:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    pub fn set_max_len(&mut self, len: usize) -> () {
```
missing `///` on `pub`, and many more down in this file ;-)

lkatalin commented on intel-sgx/attestation-daemon/src/main.rs:
```diff
@@ -0,0 +1,44 @@
+    // The attestation daemon handles each incoming connection from a tenant. The tenant, by
+    // connecting, is requesting an attestation of the enclave.
+    for incoming_tenant_stream in TcpListener::bind(LISTENER_CONN).unwrap().incoming() {
+        // The attestation daemon retrieves the Quoting Enclave's Target Info and
```
Yes, I believe so. In the Fortanix code it comes from [here](https://github.com/fortanix/rust-sgx/blob/master/dcap-ql-sys/src/lib.rs#L76), but I haven't been able to find enough detail on how Intel's `sgx_qe_get_target_info` works. I'll look into it further.

lkatalin commented on intel-sgx/Cargo.toml:
```diff
@@ -0,0 +1,6 @@
+members = ['attestation-enclave', 'attestation-daemon', 'attestation-tenant']
+
+[patch.crates-io]
+dcap-ql = { git = "https://github.com/lkatalin/rust-sgx", branch = "serde-pck" }
```
This is good to know! Thanks.

lkatalin commented:
@steveeJ I think I've addressed all of your comments - let me know what you think, when you get a chance. Thanks.

steveeJ commented on intel-sgx/attestation-enclave/src/main.rs:
```diff
@@ -0,0 +1,40 @@
+use std::net::TcpListener;
+
+const LISTENER_ADDR: &'static str = "localhost:1032";
+const SER_TARGETINFO_SIZE: usize = 196; // TODO: Hope to not use this.
```
What does this TODO mean?

steveeJ replied:
As discussed out-of-band, this number has been found through a manual process of computing the struct's size. Hardcoding this number is potentially brittle. An option would be to pass the stream to serde directly via its `from_reader` method. However @lkatalin expressed an issue with that, as it consumes then stream and can't be written to later on.

steveeJ replied:
This is fixed now.

steveeJ commented on intel-sgx/attestation-enclave/src/main.rs:
```diff
@@ -0,0 +1,40 @@
+        // TODO: I'd love to be able to read this into a Vec or some other more dynamic way
+        // that doesn't involve a buffer of hard-coded size.
+        let mut buf = [0; SER_TARGETINFO_SIZE];
+        stream.read_exact(&mut buf)?;
```
Regarding the comment above it would be possible to read to `EOF`, if the stream provides that. In any case we'd probably still want an upper limit to not hang endlessly.

lkatalin replied:
Let me know if you find a way to do this, thanks.

steveeJ replied:
I found a way, as you already know :wink:

steveeJ commented on intel-sgx/attestation-tenant/src/main.rs:
```diff
@@ -0,0 +1,115 @@
+    // tenant.
+    let cert_chain_file = env::args()
+        .nth(1)
+        .expect("You must supply the path of a valid PCK certificate chain as the first argument.");
```
I think this check should be moved to the beginning of the program to be able to fail early.

lkatalin replied:
Done.

lkatalin commented on intel-sgx/attestation-tenant/src/key.rs:
```diff
@@ -0,0 +1,65 @@
+        hasher.update(&unhashed_data);
+        let hash = hasher.finish();
+        if hash != hashed_data {
+            panic!("Hash of attestation key is not valid.")
```
I changed this to a custom `HashError`.

lkatalin commented:
@steveeJ I fixed the error handling in `attestation-tenant/key.rs` and moved the argument check to the beginning of `attestation-tenant/main.rs`.

steveeJ commented:
@lkatalin would you mind just quickly confirming that the following comments by @npmccallum are addressed, before I merge this?

> * OpenSSL (and mbedtls) should have functions for the type conversions you're doing. Please confirm.
> 
> * It is unclear to me that `verify_chain_sigs()` actually does signature validation. At least that's not how I'm used to doing signature validation. Traditionally, I've built up a full certificate chain and called a single verify function on the whole chain. For example, if Intel decided to add multiple intermediate certificates this could would have to be re-written.

lkatalin commented:
Yes, I used to have lots of manual functions converting from DER to ASN.1 and things like that, because I didn't realize OpenSSL had such functionality. Those were all addressed.

I also had a different function for verifying the chain that checked each intermediate cert in the chain. There is also a function for this where I can just check the leaf cert, which is what's used instead now in `cert_chain.rs`. It loops through the chain, so a changing number of intermediate certs won't break it. 

So, these were both addressed.
